api-put-object-copy.go

```diff
@@ -31,19 +35,47 @@ func (c Client) CopyObject(bucketName string, objectName string, objectSource st
 	if err := isValidObjectName(objectName); err != nil {
 		return err
 	}
-	if objectSource == "" {
-		return ErrInvalidArgument("Object source cannot be empty.")
+	srcBucket, srcObject, err := getObjectSource(objectSource)
+	if err != nil {
+		return err
+	}
+
+	// Get info about the source object
+	srcInfo, err := c.StatObject(srcBucket, srcObject)
+	if err != nil {
+		return err
+	}
+	srcByteRangeSize := cpCond.getRangeSize()
+	if srcByteRangeSize > srcInfo.Size ||
+		(srcByteRangeSize > 0 && cpCond.byteRangeEnd >= srcInfo.Size) {
+		return ErrInvalidArgument(fmt.Sprintf(
+			"Specified byte range (%d, %d) does not fit within source object (size = %d)",
+			cpCond.byteRangeStart, cpCond.byteRangeEnd, srcInfo.Size))
+	}
+
+	copySize := srcByteRangeSize
```

Comment:
A comment explaining how an empty range is interpreted as the entire object would provide context for future.